### PR TITLE
Port helper functions,_ldexp_exp and _frexp_exp, for hyperbolic functions

### PR DESCRIPTION
**Pull request by pkofod**
cc @simonbyrne 

So this is supposed to remove the `ccall`s to `__ldexp_exp` in the hyperbolic calls. It should work exactly like the Openlibm version, but it **will not** give the same outputs for **all** inputs. It will for most all inputs, but it differens whenever `exp(x)` differs from a call to Openlibm's `exp(x)`. We can look into this if necessary.

I've rearranged a lot of the code from https://github.com/JuliaMath/openlibm/blob/cca41bc1abd01804afa4862bbd2c79cc9803171a/src/k_exp.c#L59 and https://github.com/JuliaMath/openlibm/blob/cca41bc1abd01804afa4862bbd2c79cc9803171a/src/k_expf.c#L59 so those are the files to check if you want to check for errors, or if something in the comments and doc strings are unclear.

### CODE REVIEW

File: base/special/ldexp_exp.jl [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+	# respectively.
+
+	# We use exp(x) = exp(x - kln2) * 2**k, carefully chosen to
+    # minimize |exp(kln2) - 2**k|.
+	kr = T == Type{Float32} ? UInt32(235) : UInt32(1799)
+
+	# We also scale the exponent of exp_x to exponent_bias + the largest finite
+	# exponent (exponent of T(Inf)-1, so that the result can be multiplied by
+    # a tiny number without losing accuracy due to denormalization.
+	exp_x = exp(x - kr*log(T(2))) # exp_x*2^k = exp(x)
```
[the remark this answers is not in the document]
Comment (pkofod) [reply base/special/ldexp_exp.jl]: Increase relative to what? `x` is so large that `exp(x)` can't be evaluated, so the error would be infinitely big if we didn't do it. Anyway, this is essentially a port, so this is how `__ldexp_exp` and all the hyperbolics that call it have been working all along.

**Comment by pkofod** [comment]
I guess we can run the scalar benchmarks here, but I really doubt much can regress here (famous last words). I can't fire them though; @simonbyrne ?

**Comment by KristofferC** [comment]
@nanosoldier `runbenchmarks("scalar", vs = ":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/29499#issuecomment-428558338) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/d348a1bb38d9732cc3b855ed36c13a3a32602949/504dd67_vs_367aab3/report.md). cc @ararslan

**Comment by pkofod** [comment]
@simonbyrne besides the point that you mentioned in your review (that I'd honestly prefer to save for a later improvement just to close this, since this is meant to port the functionality to weaken the dependency on openlibm), are there any things that jumps out at you? Maybe a second nanosoldier run is in place to see what is noise in the run above.

File: base/special/ldexp_exp.jl [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+modify_highword(x::Float32, hw) = reinterpret(Float32, hw)
+modify_highword(x::Float64, hw) = reinterpret(Float64, (UInt64(hw)<<32)|(reinterpret(UInt64, x)<<32)>>32)
+
+exponent_rshift(T::Type{Float32}, hw) = hw >> 23 # this comes from 32 (bits in UInt32) minus 9 bits for the sign and exponent
+exponent_rshift(T::Type{Float64}, hw) = hw >> 20 # this comes from 32 (bits in UInt32) minus 12 bits for the sign and exponent
+exponent_lshift(T::Type{Float32}, hw) = hw << 23 # this comes from 32 (bits in UInt32) minus 9 bits for the sign and exponent
+exponent_lshift(T::Type{Float64}, hw) = hw << 20 # this comes from 32 (bits in UInt32) minus 12 bits for the sign and exponent
+
+function modify_exponent(x::T, expnt_x) where T <: Union{Float32, Float64}
+	# mask away sign and exponent; "000...0111..111" with 9 or 12 leading 0's
```
Comment (ararslan) [inline base/special/ldexp_exp.jl]: Indentation should be 4 spaces instead of tabs.

**Comment by ararslan** [comment]
@nanosoldier `runbenchmarks("scalar", vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/29499#issuecomment-448732081) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/fa323242701b9f212e7000927f6c69a4178bbd15/21aa5e6_vs_2c71488/report.md). cc @ararslan

**Comment by ararslan** [comment]
The only functions that use these functions internally are `sinh` and `cosh`, and none of the others use `sinh` or `cosh` internally, so the changes here affect only those two functions.

Here is the reduced set of Nanosoldier results for just `cosh`, as `sinh` showed no changes in this or the previous run:

| ID | time ratio | memory ratio |
|----|------------|--------------|
| `["scalar", "cosh", "(\"0 <= abs(x) < 2.7755602085408512e-17\", \"negative argument\", \"Float64\")"]` | 1.10 (5%) :x: | 1.00 (1%)  |
| `["scalar", "cosh", "(\"9f0 <= abs(x) < 88.72283f0\", \"negative argument\", \"Float32\")"]` | 1.18 (5%) :x: | 1.00 (1%)  |
| `["scalar", "cosh", "(\"9f0 <= abs(x) < 88.72283f0\", \"positive argument\", \"Float32\")"]` | 0.82 (5%) :white_check_mark: | 1.00 (1%)  |
| `["scalar", "cosh", "(\"very large\", \"negative argument\", \"Float64\")"]` | 1.15 (5%) :x: | 1.00 (1%)  |
| `["scalar", "cosh", "(\"very small\", \"negative argument\", \"Float32\")"]` | 1.10 (5%) :x: | 1.00 (1%)  |
| `["scalar", "cosh", "(\"very small\", \"positive argument\", \"Float64\")"]` | 1.10 (5%) :x: | 1.00 (1%)  |

Note that the previous run showed similar results for these particular benchmarks.

**Comment by ararslan** [comment]
Simon doesn't have time right now to go over this, and he said he trusts your and my judgement regarding performance. I agree with your assessment that it looks like noise, so I'm just going to merge it.

**Pull request merged**